Rework of PillarDAO

cryptonoyaiba opened the pull request:
- Allows for passing in at deployment an existing array of accounts (for our specific use case). This will allow for any members that already exist to be migrated onto a new DAO contract. This expects a MembershipNFT contract to already be in use with an existing set of NFTs minted to those members.
- Addition of deployment script with setup for minting new members to existing MembershipNFT contract.
- Added test to mimic entire process through to withdrawal.
- Upgraded hardhat version to 2.13.0 and new hardhat-network-helpers package for testing time dependant tests.

cryptonoyaiba committed:
Rework of PillarDAO

- Allows for passing in at deployment an existing array of accounts (for our specific use case). This will allow for any members that already exist to be migrated onto a new DAO contract. This expects a MembershipNFT contract to already be in use with an existing set of NFTs minted to those members.
- Addition of deployment script with setup for minting new members to existing MembershipNFT contract.
- Added test to mimic entire process through to withdrawal.
- Upgraded hardhat version to 2.13.0 and new hardhat-network-helpers package for testing time dependant tests.

ch4r10t33r commented on contracts/PillarDAO.sol:
```diff
@@ -127,32 +128,52 @@ contract PillarDAO is IPillarDAO, Ownable, ReentrancyGuard {
-    // }
+    function _addExistingMembers(address[9] memory _members) internal {
+        // pre-add existing DAO members [9]
+        memberships[_members[0]] = 1;
```
Should we add a validation to ensure the address is not 0?

cryptonoyaiba replied:
Added validation for `address(0)`. Also changed this from `address[9]` to a dynamic length incase we need to reuse and make `_addExistingMembers` a loop to account for dynamic length. Added test in suite to confirm revert occurs.

ch4r10t33r commented on contracts/PillarDAO.sol:
```diff
@@ -127,32 +128,52 @@ contract PillarDAO is IPillarDAO, Ownable, ReentrancyGuard {
+        memberships[_members[2]] = 3;
+        balances[_members[2]] = Deposit({
+            depositAmount: 10000 ether,
+            depositTime: block.timestamp
```
Shouldn't we retain the original timestamp of the membership?

cryptonoyaiba replied:
@ch4r10t33r The only way I could do this is look when the `deposit` tx went through and convert that to a UNIX timestamp. There is no way to view that info from the contract. That would require me to change the shape of the data passed in. I can do this if you are happy with that approach?

ch4r10t33r replied:
Cant we pass the timestamps also as an additional array?

ch4r10t33r commented on package.json:
```diff
@@ -1,13 +1,14 @@
 {
   "name": "pillar-dao-contracts",
   "devDependencies": {
+    "@nomicfoundation/hardhat-network-helpers": "^1.0.9",
```
Pls stick to a specific version of the package.

cryptonoyaiba replied:
done

cryptonoyaiba committed:
PillarDAO contract changes

- Changed fixed length `_preExistingMembers` array to be dynamic.
- Added functionality to `setDepositTimestamp` to allow for adding timestamps for pre existing members.
- Added to test suite.
- Created new deploy script with configuration for DAO migration.

cryptonoyaiba commented:
Set up test environment for QA (migration completed) on Mumbai.

cryptonoyaiba committed:
Deployed new version to Polygon

- Migrated users from incorrect `MembershipNFT` contract to correct one.
- Set new `PillarDAO` as `vault` in correct `MembershipNFT` contract.
- Added deposit timestamps for existing users.
- Updated `DEPLOYMENTS.md`.
